toolchain/lower/mangler.cpp

```diff
@@ -0,0 +1,91 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/lower/mangler.h"
+
+#include "toolchain/base/kind_switch.h"
+#include "toolchain/sem_ir/entry_point.h"
+
+namespace Carbon::Lower {
+
+auto Mangler::MangleInverseQualifiedNameScope(bool first_name_component,
+                                              llvm::raw_ostream& os,
+                                              SemIR::NameScopeId name_scope_id)
+    -> void {
+  auto& sem_ir = file_context_.sem_ir();
+  auto names = sem_ir.names();
+  auto types = sem_ir.types();
+  while (name_scope_id.is_valid()) {
+    const auto& parent = sem_ir.name_scopes().Get(name_scope_id);
+    if (parent.inst_id == SemIR::InstId::PackageNamespace) {
+      break;
+    }
```

Comment:
I think this is similar and avoids a lookup for the package namespace:

```suggestion
  while (name_scope_id.is_valid() && name_scope_id != SemIR::NameScopeId::Package) {
    const auto& parent = sem_ir.name_scopes().Get(name_scope_id);
```